Adding lazy loading support for datastore connection.

dhermes opened the pull request:
**NOTE**: Has #667 as diffbase.

Happens in 4 phases:

- Remove `datastore.helpers` import in `connection`. (this would have caused a cycle if `Connection` was used in `_implicit_environ`)
- Moving `get_connection()` from `datastore.__init__` to `_implicit_environ`.
- Moving `set_default_connection()` from `datastore.__init__` to `_implicit_environ`.
- Adding lazy loading support for the connection property

Note that I had to copy `_prepare_key_for_request` from `helpers` into `connection` to avoid cyclic imports (#528 would be nice, since we could just remove `_prepare_key_for_request` all together)

tseaver commented on gcloud/datastore/_implicit_environ.py:
```diff
@@ -96,4 +156,102 @@ def get_default_dataset_id():
+    :param name: The name of the attribute / property being evaluated.
+
+    :type method: callable that takes no arguments
+    :param method: The method used to evaluate the property.
```
`_method` seems misleading here:  for instance, `_lazy_dataset_id` is a free function, not a method.  Maybe `_deferred` or `_deferred_callable` instead?

tseaver commented:
The last commit LGTM, except for my note about the `_method` name.

dhermes commented:
@tseaver PTAL I rebased on top of #667.

Your `_method` comment was resolved in #667. You gave the LGTM for the final commit, but this PR actually involved 3 commits to move code before (on of which involves a duplicated method and `pragma NO COVER` worth discussing)

Also the commit you LGTM'd changed slightly since I changed #667 to use a decorator for lazy properties

tseaver commented on gcloud/datastore/batch.py:
```diff
@@ -189,8 +189,8 @@ def delete(self, key):
 
-        key_pb = key.to_protobuf()
-        helpers._add_keys_to_request(self.mutation.delete, [key_pb])
+        key_pb = helpers._prepare_key_for_request(key.to_protobuf())
```
What does this change have to do with this PR?  And I though `_prepare_key_for_request` was dead?

dhermes replied:
I'm on mobile now so can't write much. Read the PR description for an explanation of the purpose for each of the 4 commits.

tseaver replied:
This change isn't about removing a `datastore.helpers` import:  it changes how the deleted key gets added to the mutation (effectively, in inlines `helpers._add_keys_to_request`).

dhermes replied:
`_add_keys_to_request` was only used here and in the place it is newly defined (in `connection`). `_add_keys_to_request` was added to be used with lists, so for one `key_pb` here `CopyFrom` suffices.

Removing the `datastore.helpers` in `connection` was made easier by moving the definition of `_add_keys_to_request` there.

tseaver commented:
I've got to run, won't be able to finish review tonight. :(

dhermes commented:
No worries

dhermes commented:
@tseaver After this is in I'd like to cut `0.4.2`. WDYT?

tseaver commented:
OK, LGTM.  0.4.2 would be fine, as well.
